File: leetcode-java/src/main/java/chapter02/ReverseList.java
package chapter02;

import utils.ListNode;

/**
 * <p>反转链表</p>
 * <p>1.反转链表</p>
 * <p>2.反转链表 II</p>
 * <p>3.两两交换链表中的结点</p>
 * <p>4.K个一组反转</p>
 * <p>万能解法: 集合、数组、栈都是可以完成的</p>
 */
public class ReverseList {

    /**
     * <p>思路: 反转链表 迭代</p>
     */
    public static ListNode reverseListLoop(ListNode head){
        ListNode previous = null;
        ListNode current = head;
        ListNode next = null;
        while (current != null) {
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        return previous;
    }

    /**
     * <p>思路: 反转链表 递归</p>
     */
    public static ListNode reverseListRecursive(ListNode head){
        if (head == null || head.next == null) {
            return head;
        }
        ListNode tail = reverseListRecursive(head.next);
        head.next.next = head;
        head.next = null;
        return tail;
    }

    /**
     * <p>思路: 反转链表 II</p>
     */
    public static ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode(0, head);
        ListNode current = dummy, start = null, end = null;
        ListNode leftNode = null, rightNode = null;
        int index = 0;
        for (index = 0; index < left - 1; index++) {
            current = current.next;
        }
        leftNode = current;
        for (index = left; index < right + 1; index++) {
            current = current.next;
        }
        rightNode = current;
        start = leftNode.next;
        end = rightNode.next;
        leftNode.next = null;
        rightNode.next = null;
        reverseListLoop(start);
        leftNode.next = rightNode;
        start.next = end;
        return dummy.next;
    }

    /**
     * <p>思路: K个一组反转链表 <=> 两两交换链表中的结点</p>
     */
    public static ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0 ,head);
        ListNode current = head, previous = dummy, start = null, end = null;
        while (current != null) {
            for (int index = 0; index < k - 1; index++) {
                current = current.next;
                if (current == null) {
                    return dummy.next;
                }
            }
            start = previous.next;
            end = current.next;
            previous.next = null;
            current.next = null;
            reverseListLoop(start);
            previous.next = current;
            start.next = end;
            previous = start;
            current = end;
        }
        return dummy.next;
    }
}

File: leetcode-go/chapter02/reverse_list.go
package main

type ListReverser struct {
}

// 反转链表: 迭代
func (reverse *ListReverser) ReverseListLoop(head *ListNode) *ListNode {
	var previous *ListNode = nil
	var current *ListNode = head
	var next *ListNode = nil
	for current != nil {
		next = current.Next
		current.Next = previous
		previous = current
		current = next
	}
	return previous
}

// 反转链表: 递归
func (reverse *ListReverser) ReverseListRecursive(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	tail := reverse.ReverseListRecursive(head.Next)
	head.Next.Next = head
	head.Next = nil
	return tail
}

// 反转链表 II
func (reverse *ListReverser) ReverseBetween(head *ListNode, left, right int) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	current := dummy
	for index := 0; index < left - 1; index++ {
		current = current.Next
	}
	leftNode := current
	for index := left - 1; index < right; index++ {
		current = current.Next
	}
	rightNode := current
	startNode := leftNode.Next
	endNode := rightNode.Next
	leftNode.Next = nil
	rightNode.Next = nil
	reverse.ReverseListLoop(startNode)
	leftNode.Next = rightNode
	startNode.Next = endNode
	return dummy.Next
}

// 反转链表: K 个一组
func (reverse *ListReverser) ReverseKGroup(head *ListNode, k int) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	current, previous := head, dummy
	for current != nil {
		// 注意: 移动的单位只能是 k - 1
		for index := 0; index < k - 1; index++ {
			if current.Next == nil {
				return dummy.Next
			}
			current = current.Next
		}
		start := previous.Next
		end := current.Next
		previous.Next = nil
		current.Next = nil
		reverse.ReverseListLoop(start)
		previous.Next = current
		start.Next = end
		previous = start
		current = end
	}
	return dummy.Next
}

// 反转链表: 两两交换
func (reverse *ListReverser) SwapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	current := dummy
	for current.Next != nil && current.Next.Next != nil {
		node1, node2 := current.Next, current.Next.Next
		current.Next = node2
		node1.Next = node2.Next
		node2.Next = node1
		current = node2.Next
	}
	return dummy.Next
}
